src/Compilers/CSharp/Test/Emit2/Semantics/InlineArrayTests.cs

```diff
@@ -13845,5 +14167,2885 @@ struct Buffer
+  IL_0006:  stloc.0
+  IL_0007:  ldloca.s   V_0
+  IL_0009:  stloc.1
+  IL_000a:  ldc.i4.0
+  IL_000b:  stloc.2
+  IL_000c:  br.s       IL_003b
+  IL_000e:  ldloc.1
+  IL_000f:  ldloc.2
+  IL_0010:  call       ""InlineArrayElementRefReadOnly<Buffer4<int>, int>(in Buffer4<int>, int)""
+  IL_0015:  ldind.i4
+  IL_0016:  ldc.i4.s   32
+  IL_0018:  call       ""void System.Console.Write(char)""
+  IL_001d:  call       ""void System.Console.Write(int)""
+  IL_0022:  ldloca.s   V_3
+  IL_0024:  dup
+  IL_0025:  initobj    ""Buffer4<int>""
+  IL_002b:  constrained. ""Buffer4<int>""
+  IL_0031:  callvirt   ""string object.ToString()""
+  IL_0036:  pop
+  IL_0037:  ldloc.2
+  IL_0038:  ldc.i4.1
+  IL_0039:  add
+  IL_003a:  stloc.2
+  IL_003b:  ldloc.2
+  IL_003c:  ldc.i4.4
+  IL_003d:  blt.s      IL_000e
+  IL_003f:  ret
+}
+");
+        }
+
+        [Fact]
+        public void Foreach_Value_04()
+        {
+            var src = @"
+class Program
+{
+    static ref int Test()
+    {
+        foreach (var y in GetBuffer())
+        {
+            return ref y;
+        }
+
+        throw null;
+    }
+
+    static Buffer4<int> GetBuffer() => default;
+}
+";
+            var comp = CreateCompilation(src + Buffer4Definition, targetFramework: TargetFramework.Net80, options: TestOptions.ReleaseDll);
+            comp.VerifyDiagnostics(
+                // (8,24): error CS1657: Cannot use 'y' as a ref or out value because it is a 'foreach iteration variable'
+                //             return ref y;
+                Diagnostic(ErrorCode.ERR_RefReadonlyLocalCause, "y").WithArguments("y", "foreach iteration variable").WithLocation(8, 24)
+                );
+        }
+
+        [Fact]
+        public void AwaitForeach_01()
+        {
+            var src = @"
+class Program
+{
+    public static async void M()
+    {
+        await foreach(var s in GetBuffer())
+        {
+        }
+    }
+
+    static Buffer4<int> GetBuffer() => default;
+}
+";
+            var comp = CreateCompilation(src + Buffer4Definition, targetFramework: TargetFramework.Net80, options: TestOptions.ReleaseDll);
+            comp.VerifyDiagnostics(
+                // (6,32): error CS8415: Asynchronous foreach statement cannot operate on variables of type 'Buffer4<int>' because 'Buffer4<int>' does not contain a public instance or extension definition for 'GetAsyncEnumerator'. Did you mean 'foreach' rather than 'await foreach'?
+                //         await foreach(var s in GetBuffer())
+                Diagnostic(ErrorCode.ERR_AwaitForEachMissingMemberWrongAsync, "GetBuffer()").WithArguments("Buffer4<int>", "GetAsyncEnumerator").WithLocation(6, 32)
+                );
+        }
+
+        [Fact]
+        public void Foreach_Extension_01()
+        {
+            var src = @"
+class Program
+{
+    public static void M(Buffer4<int> x)
+    {
+        foreach(var s in x)
+        {
+        }
+    }
+}
+
+namespace System
+{
+    public readonly ref struct Span<T>
```

Comment:
Consider adding a test with a `Span<T>` where `T` has some constraints.

Reply:
> Consider adding a test with a Span<T> where T has some constraints.

The ```Foreach_UnsupportedElementType``` unit-test already covers constraint verification. Every generic type has implicit constraints. I do not think it is worth testing against a malformed Span type.